src/cpp/session/SessionConsoleProcess.cpp

```diff
@@ -764,8 +816,36 @@ boost::shared_ptr<ConsoleProcess> ConsoleProcess::create(
 // previously used handle
 boost::shared_ptr<ConsoleProcess> ConsoleProcess::createTerminalProcess(
       core::system::ProcessOptions options,
-      boost::shared_ptr<ConsoleProcessInfo> procInfo)
+      boost::shared_ptr<ConsoleProcessInfo> procInfo,
+      bool enableWebsockets)
 {
+   boost::shared_ptr<ConsoleProcess> cp;
+
+   // Use websocket as preferred communication channel; it can fail
+   // here if unable to establish the server-side of things, in which case
+   // fallback to using Rpc.
+   //
+   // It can also fail later when client tries to connect; fallback for that
+   // happens from the client-side via RPC call procUseRpc.
+   if (enableWebsockets)
+   {
+      Error error = s_terminalSocket.ensureServerRunning();
+      if (error)
+      {
+         procInfo->setChannelMode(Rpc, "");
+         LOG_ERROR(error);
+      }
+      else
+      {
+         std::string port = boost::lexical_cast<std::string>(s_terminalSocket.port());
```

Comment:
`lexical_cast` can throw an exception which we're not prepared to handle here; use`safe_convert::numberToString`.